### PR TITLE
add Groq client support

### PR DESCRIPTION
**Pull request by ricklamers**
- Added GroqModel class to interact with Groq API.
- Implemented structured and text-based responses for Groq models.
- Added example and tests for Groq model integration.
- Updated dependencies to include groq package.

### CODE REVIEW

**Comment by ricklamers** [comment]
One thing I'm not sure of is how to let users override temperature, I've defaulted to 0.0 here because that's what most open source models need to adhere to function calling requests

**Comment by ricklamers** [comment]
Not sure how to fix CI for

```
0s
Run make docs-insiders
Error: PPPR_TOKEN is not set, can't install insiders packages
make: *** [Makefile:80: .docs-insiders-install] Error [1](https://github.com/pydantic/pydantic-ai/actions/runs/11975670356/job/33389567928?pr=84#step:6:1)
Error: Process completed with exit code 2.
```

**Comment by samuelcolvin** [comment]
This is awesome, thank you.

CI is my problem, I'll fix that — you're the first person contributing form outside pydantic (from a fork), hence you're seeing the issues.

I'm flat out on calls today, but I'll review ASAP.

File: pyproject.toml [hunk 1]
```diff
@@ -39,6 +39,7 @@ dependencies = [
     "httpx>=0.27.2",
     "logfire-api>=1.2.0",
     "openai>=1.54.3",
+    "groq>=0.12.0",
```
Comment (samuelcolvin) [inline pyproject.toml]: we should probably make all models (including OpenAI) optional groups.

File: tests/test_live.py [hunk 2]
```diff
@@ -90,3 +91,34 @@ async def test_gemini_structured(http_client: httpx.AsyncClient):
     print('Gemini structured cost:', result.cost())
     cost = result.cost()
     assert cost.total_tokens is not None and cost.total_tokens > 0
+
+
+async def test_groq(http_client: httpx.AsyncClient):
```
Comment (samuelcolvin) [inline tests/test_live.py]: I guess I'll need to set an API key for this?

I can look it up, but is the process trivial?

File: pydantic_ai/models/groq.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+from collections.abc import AsyncIterator, Iterable, Mapping, Sequence
+from contextlib import asynccontextmanager
+from dataclasses import dataclass, field
+from datetime import datetime, timezone
+from typing import Literal, overload
+
+from groq import NOT_GIVEN, AsyncGroq, AsyncStream
+from groq.types import chat
+from groq.types.chat import ChatCompletion, ChatCompletionChunk
+from groq.types.chat.chat_completion_chunk import ChoiceDeltaToolCall
```
Comment (samuelcolvin) [inline pydantic_ai/models/groq.py:12]: this looks very close to OpenAI's SDK, could we somehow reuse the OpenAI model and just change the endpoint?

File: tests/test_live.py (hunk 2 above)
Comment (ricklamers) [reply tests/test_live.py]: Free API key can be created, trivial just go to https://console.groq.com/

File: pydantic_ai/models/groq.py (hunk 3 above)
Comment (ricklamers) [reply pydantic_ai/models/groq.py:12]: There are slight differences which is why we need our client (especially in the token use x_groq parts)

File: pyproject.toml (hunk 1 above)
Comment (ricklamers) [reply pyproject.toml]: Probably a different PR, but agreed

**Comment by samuelcolvin** [comment]
As of #87, PRs should work from forks. You'll need to rebase or merge with main.

File: tests/test_live.py (hunk 2 above)
Comment (samuelcolvin) [reply tests/test_live.py]: I'll fix this once the initial PR is merged, `test-live` will no longer run on PRs from forks once you rebase.

File: pyproject.toml (hunk 1 above)
Comment (samuelcolvin) [reply pyproject.toml]: adding the group should be as simple as

```bash
uv add --optional groq groq
```

**Comment by samuelcolvin** [comment]
amazing, thank you so much.

**Comment by samuelcolvin** [comment]
Thanks so much for this @ricklamers.

Related to the documentation: it would be really nice if Groq published an `objects.inv` file for their python SDK, so we can link directly to entities in the package, as we do with other packages:

https://github.com/pydantic/pydantic-ai/blob/d29d5e0d8757179036273351acc7f31c48bebe49/mkdocs.yml#L155-L160

I asked @rattrayalex about this in relation to the OpenAI SDK, and hi said it wasn't an immediate priority, but I can't helping thinking that it wouldn't that hard for SDKs built with stainless to generate API docs with [mkdocstrings](https://mkdocstrings.github.io/) and get a `objects.inv` definition for free.

**Comment by rattrayalex** [comment]
Hey @samuelcolvin , sorry I don't recall that convo?

@robertcraigie would be the one to take this on – we _could_ generate a `objects.inv` definition, but not 100% straightforward and also not sure were it should go?

**Comment by RobertCraigie** [comment]
would providing an `objects.inv` file be helpful without also hosting docs?

**Pull request merged**